extensions/notebook/src/book/bookTrustManager.ts

```diff
@@ -0,0 +1,146 @@
+	rootPath: string;
+	getConfiguration(section?: string | undefined): vscode.WorkspaceConfiguration;
+}
+
+export class BookTrustManager implements IBookTrustManager {
+
+	private static notebookConfiguration: string = 'notebook';
+	private static notebookTrustedBooksConfiguration: string = 'trustedBooks';
+	private trustedLocalBooks: Record<string, boolean> = {};
+
+	constructor(private books: BookModel[], private workspaceDetails?: IBookTrustManagerWorkspaceDetails) {
+		if (!workspaceDetails) {
+			this.workspaceDetails = {
+				get getConfiguration() {
+					return vscode.workspace.getConfiguration;
+				},
+				get rootPath() {
+					return vscode.workspace.rootPath;
+				}
+			};
+		}
+	}
+
+	isNotebookTrustedByDefault(notebookUri: string): boolean {
+		let normalizedBookDirectory = this.getBookDirectory(notebookUri);
+		let trustedBookDirectory = this.getTrustedBookDirectory(notebookUri, normalizedBookDirectory);
+
+		if (normalizedBookDirectory && trustedBookDirectory) {
+			let trustedBook = this.getTrustedBook(normalizedBookDirectory, trustedBookDirectory);
+
+			if (trustedBook) {
+				let fullBookBaseUri = trustedBookDirectory !== normalizedBookDirectory
+					? path.join(normalizedBookDirectory, trustedBookDirectory) : normalizedBookDirectory;
+				let fullBookBaseUriWithContent = path.join(fullBookBaseUri, 'content');
+				let requestingNotebookFormattedUri = notebookUri.substring(fullBookBaseUriWithContent.length).replace('.ipynb', '');
+				let notebookInTOC = trustedBook.tableOfContents.sections.find(jupyterSection => {
+					let normalizedJupyterSectionUrl = jupyterSection.url && path.normalize(jupyterSection.url);
+					return normalizedJupyterSectionUrl === requestingNotebookFormattedUri;
+				});
+				return !!notebookInTOC;
+			}
+		}
+
+		return false;
+	}
+
+	setBookAsTrusted(bookRootPath: string): boolean {
+		// add this TOC to the configuration list
+		let workspacePathLength: number = this.workspaceDetails.rootPath ? path.resolve(this.workspaceDetails.rootPath).length : 0;
+		let relativeBookPathStartingIndex = !!workspacePathLength ? workspacePathLength + 1 : 0;
+		let relativeBookPath: string = path.normalize(path.resolve(bookRootPath).substring(relativeBookPathStartingIndex));
+		let existingBooks: string[] = this.getTrustedBooksInConfig();
+
+		// if no match found in the configuration, then add it
+		if (!existingBooks.find(notebookPath => path.normalize(notebookPath) === relativeBookPath)) {
+			existingBooks.push(relativeBookPath);
+
+			// update the configuration
+			this.setTrustedBooksInConfig(existingBooks);
+			return true;
+		}
+		return false;
+	}
+
+	getBookDirectory(notebookUri: string): string {
+		let workspace: vscode.WorkspaceFolder = this.getNotebookWorkspaceFolder(notebookUri);
+		let normalizedContainerPath: string = workspace ? path.normalize(workspace?.uri.fsPath) : undefined;
+
+		if (!normalizedContainerPath) {
+			normalizedContainerPath = this.books
```

Comment:
So if I'm understanding this correctly this is where you'll look through all open books if you didn't find it in the open workspaces? A couple questions :

1. Why wouldn't you just look in the open books directly? That should contain all books right? (IIRC we load all books in a workspace when that workspace is opened)
2. Did you look and see if there's a helper of some sort to see if a book is currently open? (basically this logic but in a more centralized place). If not I would suggest adding something like that since this seems like a general case that we may want in the future

Reply:
Ah yes, you make a very good point. Initially the design was to trust the directory but it has shifted to the notebooks in the table of contents. I have shifted the design to leverage the books instead of directories. Overall, I think it's more readable. Thanks for pointing it out.